### PR TITLE
FPLA-559: Provide statement of service (c9)

### PR DESCRIPTION
**Pull request by rebeccabrennan**
### JIRA link (if applicable) ###
https://tools.hmcts.net/jira/browse/FPLA-559

### Change description ###
Creating form for c9 to enter recipient details.

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[ x] No
```

### CODE REVIEW

File: ccd-definition/CaseEventToFields/addStatementOfService.json [hunk 1]
```diff
@@ -0,0 +1,47 @@
+[
+  {
+    "LiveFrom": "01/01/2017",
+    "CaseTypeID": "CARE_SUPERVISION_EPO",
+    "CaseEventID": "addStatementOfService",
+    "CaseFieldID": "c9Declaration",
```
Comment (Jamesnelson117) [inline ccd-definition/CaseEventToFields/addStatementOfService.json]: Should this not be statementOfServiceDeclaration to be consistent with the serviceDeclarationLabel?

File: ccd-definition/ComplexTypes/Recipients.json [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+    "ElementLabel": "Time sent",
+    "HintText": "For example, 2:30pm",
+    "SecurityClassification": "Public"
+  },
+  {
+    "LiveFrom": "01/01/2017",
+    "ID": "Recipients",
+    "ListElementCode": "sentBy",
+    "FieldType": "FixedRadioList",
+    "FieldTypeParameter": "StatementOfServiceSentBy",
```
Comment (Jamesnelson117) [inline ccd-definition/ComplexTypes/Recipients.json]: Do we need the 'StatementOfService' text here? Makes it less reusable.

File: ccd-definition/FixedLists.json [hunk 3]
```diff
@@ -597,5 +597,26 @@
     "ListElementCode": "NONE",
     "ListElement": "None",
     "DisplayOrder": 7
+  },
+  {
+    "LiveFrom": "01/01/2017",
+    "ID": "StatementOfServiceSentBy",
+    "ListElementCode": "EMAIL",
+    "ListElement": "email",
```
Comment (Jamesnelson117) [inline ccd-definition/FixedLists.json]: Do we not need title case on each of the three listElement?

File: codecept.conf.js [hunk 4]
```diff
@@ -57,6 +57,7 @@ exports.config = {
     uploadStandardDirectionsDocumentEventPage: './e2e/pages/events/uploadStandardDirectionsDocumentEvent.page.js',
     sendCaseToGatekeeperEventPage: './e2e/pages/events/sendCaseToGatekeeperEvent.page.js',
     addHearingBookingDetailsEventPage: './e2e/pages/events/addHearingBookingDetailsEvent.page.js',
+    provideStatementOfServiceEventPage: './e2e/pages/events/provideStatementOfServiceEvent.page.js',
```
Comment (Jamesnelson117) [inline codecept.conf.js]: addStatmentOfServiceEventPage* to match the event name

File: e2e/pages/events/provideStatementOfServiceEvent.page.js [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+  },
+
+  addRecipientButton: '//*[@id="statementOfService"]/div/button[1]',
+
+  addRecipient() {
+    I.click(this.addRecipientButton);
+    I.wait(1);
+  },
+
+  async enterRecipientDetails(recipients) {
```
Comment (Jamesnelson117) [inline e2e/pages/events/provideStatementOfServiceEvent.page.js]: Can we split this method up into smaller more focus methods e.g enterReceipentsAddress etc

File: e2e/pages/events/provideStatementOfServiceEvent.page.js [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+    I.waitForText('Recipient\'s email address');
+    within(this.fields(elementIndex).recipients.sentBy.email, () => {
+      I.fillField(this.fields(elementIndex).recipients.emailAddress, 'email@email.com');
+    },);
+  },
+
+  giveDeclaration() {
+    I.checkOption('I agree with this statement');
+  },
+
```
Comment (Jamesnelson117) [inline e2e/pages/events/provideStatementOfServiceEvent.page.js]: Watch out for un-needed whitespace.

File: e2e/tests/localAuthorityMaintainsCaseAfterSubmission_test.js [hunk 7]
```diff
@@ -50,3 +51,41 @@ Scenario('local authority uploads court bundle', (I, uploadDocumentsEventPage, s
...
+  caseViewPage.selectTab(caseViewPage.tabs.legalBasis);
+  I.seeAnswerInTab(1, 'Recipients 1', 'Name of recipient', recipients[0].name);
+  I.seeAnswerInTab(2, 'Recipients 1', 'Do you have the recipient\'s address?', recipients[0].addressCheck);
+  I.seeAnswerInTab(1, 'Recipient\'s address', 'Building and Street', recipients[0].address.buildingAndStreet.lineOne);
+  I.seeAnswerInTab(2, 'Recipient\'s address', 'Address Line 2', recipients[0].address.buildingAndStreet.lineTwo);
+  I.seeAnswerInTab(3, 'Recipient\'s address', 'Address Line 3', recipients[0].address.buildingAndStreet.lineThree);
+  I.seeAnswerInTab(4, 'Recipient\'s address', 'Town or City', recipients[0].address.town);
+  I.seeAnswerInTab(5, 'Recipient\'s address', 'Postcode/Zipcode', recipients[0].address.postcode);
+  I.seeAnswerInTab(6, 'Recipient\'s address', 'Country', recipients[0].address.country);
+  I.seeAnswerInTab(4, 'Recipients 1', 'Documents', recipients[0].documents);
```
Comment (Jamesnelson117) [inline e2e/tests/localAuthorityMaintainsCaseAfterSubmission_test.js]: What happened to 3?

File: service/src/main/java/uk/gov/hmcts/reform/fpl/controllers/StatementOfServiceController.java [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+@RequestMapping("/callback/statement-of-service")
+public class StatementOfServiceController {
+
+    private static final String CONSENT_TEMPLATE = "I, %s, have served the documents as stated.";
+    private final UserDetailsService userDetailsService;
+    private final MapperService mapperService;
+    private final StatementOfServiceService statementOfServiceService;
+
+    @Autowired
+    public StatementOfServiceController(UserDetailsService userDetailsService,MapperService mapperService,
```
Comment (Jamesnelson117) [inline service/src/main/java/uk/gov/hmcts/reform/fpl/controllers/StatementOfServiceController.java]: Break into separate lines

File: service/src/main/java/uk/gov/hmcts/reform/fpl/model/Recipients.java [hunk 9]
```diff
@@ -0,0 +1,23 @@
...
+import lombok.AllArgsConstructor;
+import lombok.Builder;
+import lombok.Data;
+
+import java.time.LocalDate;
+
+@Data
+@Builder
+@AllArgsConstructor
+public class Recipients {
```
Comment (Jamesnelson117) [inline service/src/main/java/uk/gov/hmcts/reform/fpl/model/Recipients.java]: Pitty that we can't use Reform party here :/

File: ccd-definition/ComplexTypes/Recipients.json [hunk 10]
```diff
@@ -0,0 +1,93 @@
...
+    "FieldType": "AddressUK",
+    "ElementLabel": "Recipient's address",
+    "HintText": "Enter a UK postcode",
+    "FieldShowCondition": "addressCheck=\"Yes\"",
+    "SecurityClassification": "Public"
+  },
+  {
+    "LiveFrom": "01/01/2017",
+    "ID": "Recipients",
+    "ListElementCode": "documents",
```
Comment (Jamesnelson117) [inline ccd-definition/ComplexTypes/Recipients.json]: Can we be more specific with the naming here?

File: service/src/test/java/uk/gov/hmcts/reform/fpl/service/StatementOfServiceServiceTest.java [hunk 11]
```diff
@@ -0,0 +1,45 @@
...
+        assertThat(alteredHearingList.size()).isEqualTo(1);
+    }
+
+    @Test
+    void shouldReturnRecipientsIfRecipientsIsPrePopulated() {
+        CaseData caseData = CaseData.builder()
+            .statementOfService(
+                ImmutableList.of(Element.<Recipients>builder()
+                    .value(
+                        Recipients.builder().name("Recipient name").build()
```
Comment (Jamesnelson117) [inline service/src/test/java/uk/gov/hmcts/reform/fpl/service/StatementOfServiceServiceTest.java]: You could get this all onto one line

File: service/src/integrationTest/java/uk/gov/hmcts/reform/fpl/controllers/StatementOfServiceControllerTest.java [hunk 12]
```diff
@@ -0,0 +1,66 @@
...
+                .header("authorization", AUTH_TOKEN)
+                .contentType(MediaType.APPLICATION_JSON)
+                .content(MAPPER.writeValueAsString(request)))
+            .andExpect(status().isOk())
+            .andReturn();
+
+        AboutToStartOrSubmitCallbackResponse callbackResponse = MAPPER.readValue(response.getResponse()
+            .getContentAsByteArray(), AboutToStartOrSubmitCallbackResponse.class);
+
+        assertThat(callbackResponse.getData()).containsKey("statementOfService");
```
Comment (Jamesnelson117) [inline service/src/integrationTest/java/uk/gov/hmcts/reform/fpl/controllers/StatementOfServiceControllerTest.java]: Can you also assert the data you expect please.

File: e2e/pages/events/provideStatementOfServiceEvent.page.js [hunk 13]
```diff
@@ -0,0 +1,70 @@
...
+  addRecipientButton: '//*[@id="statementOfService"]/div/button[1]',
+
+  addRecipient() {
+    I.click(this.addRecipientButton);
+    I.wait(1);
+  },
+
+  async enterRecipientDetails(recipients) {
+    const elementIndex = await this.getActiveElementIndex();
```
Comment (Jamesnelson117) [inline e2e/pages/events/provideStatementOfServiceEvent.page.js]: Why do we need this?

File: e2e/config.js [hunk 14]
```diff
@@ -64,6 +64,7 @@ module.exports = {
     amendOtherProceedings: 'Other proceedings',
     amendAttendingHearing: 'Attending the hearing',
     addHearingBookingDetails: 'Add hearing details',
+    provideStatementOfService: 'Add statement of service (c9)',
```
Comment (Jamesnelson117) [inline e2e/config.js]: I think it should be addStatmentOfServiceEventPage

File: ccd-definition/CaseEventToFields/addStatementOfService.json (hunk 1 above)
Comment (rebeccabrennan) [reply ccd-definition/CaseEventToFields/addStatementOfService.json]: I think the event is to add the statement of service and not just the declaration and so I think it should remain addStatementOfService as addStatementOfServiceDeclaration makes me think we are only submitting the declaration and not the event.

File: ccd-definition/ComplexTypes/Recipients.json (hunk 2 above)
Comment (rebeccabrennan) [reply ccd-definition/ComplexTypes/Recipients.json]: good shout :)

File: ccd-definition/FixedLists.json (hunk 3 above)
Comment (rebeccabrennan) [reply ccd-definition/FixedLists.json]: Sam asked for that before he wanted all list elements to be declared as enums.

File: codecept.conf.js (hunk 4 above)
Comment (rebeccabrennan) [reply codecept.conf.js]: 👍

File: e2e/pages/events/provideStatementOfServiceEvent.page.js (hunk 5 above)
Comment (rebeccabrennan) [reply e2e/pages/events/provideStatementOfServiceEvent.page.js]: I think the form is small enough it is not worth to split up and is better to keep it as one flowing form.

File: e2e/pages/events/provideStatementOfServiceEvent.page.js (hunk 6 above)
Comment (rebeccabrennan) [reply e2e/pages/events/provideStatementOfServiceEvent.page.js]: 👀

File: ccd-definition/FixedLists.json (hunk 3 above)
Comment (Jamesnelson117) [reply ccd-definition/FixedLists.json]: I mean on the listElements, should email not be Email etc

File: e2e/tests/localAuthorityMaintainsCaseAfterSubmission_test.js (hunk 7 above)
Comment (rebeccabrennan) [reply e2e/tests/localAuthorityMaintainsCaseAfterSubmission_test.js]: So it sees the recipients address as the third element and then when it goes back to recipients 1 it is the next in the list so it is 4

File: service/src/main/java/uk/gov/hmcts/reform/fpl/model/Recipients.java (hunk 9 above)
Comment (rebeccabrennan) [reply service/src/main/java/uk/gov/hmcts/reform/fpl/model/Recipients.java]: I know :(

File: ccd-definition/ComplexTypes/Recipients.json (hunk 10 above)
Comment (rebeccabrennan) [reply ccd-definition/ComplexTypes/Recipients.json]: Discussed and keeping as is as the text box description is very generic and has the ability to type anything about the documents.

File: service/src/test/java/uk/gov/hmcts/reform/fpl/service/StatementOfServiceServiceTest.java (hunk 11 above)
Comment (rebeccabrennan) [reply service/src/test/java/uk/gov/hmcts/reform/fpl/service/StatementOfServiceServiceTest.java]: 👍

File: service/src/integrationTest/java/uk/gov/hmcts/reform/fpl/controllers/StatementOfServiceControllerTest.java (hunk 12 above)
Comment (rebeccabrennan) [reply service/src/integrationTest/java/uk/gov/hmcts/reform/fpl/controllers/StatementOfServiceControllerTest.java]: 👍

File: e2e/pages/events/provideStatementOfServiceEvent.page.js (hunk 13 above)
Comment (rebeccabrennan) [reply e2e/pages/events/provideStatementOfServiceEvent.page.js]: it is consistent with what we have in other places :) discussed and resolved.

File: e2e/config.js (hunk 14 above)
Comment (rebeccabrennan) [reply e2e/config.js]: 👍 changed

File: e2e/tests/localAuthorityMaintainsCaseAfterSubmission_test.js [hunk 15]
```diff
@@ -50,3 +51,41 @@ Scenario('local authority uploads court bundle', (I, uploadDocumentsEventPage, s
   caseViewPage.selectTab(caseViewPage.tabs.documents);
   I.seeDocument('Court bundle', 'mockFile.txt');
 });
+
+Scenario('local authority provides a statements of service', async (I, caseViewPage, loginPage, addStatementOfServiceEventPage) => {
+  caseViewPage.goToNewActions(config.administrationActions.addStatementOfService);
+  await addStatementOfServiceEventPage.enterRecipientDetails(recipients[0]);
```
Comment (akomisarek) [inline e2e/tests/localAuthorityMaintainsCaseAfterSubmission_test.js]: Do we need those awaits here? Not sure why they would be require - we should enter the data and should be able to progress 🤔

File: service/src/main/java/uk/gov/hmcts/reform/fpl/model/CaseData.java [hunk 16]
```diff
@@ -79,4 +79,5 @@
     @Valid
     private final List<@NotNull(message = "You need to add details to children")Element<Child>> children1;
     private final List<Element<HearingBooking>> hearingDetails;
+    private final List<Element<Recipients>> statementOfService;
```
Comment (akomisarek) [inline service/src/main/java/uk/gov/hmcts/reform/fpl/model/CaseData.java]: Personally I think the name of variable confusing - we have got type of recipients but the field is statementOfService?

File: service/src/main/java/uk/gov/hmcts/reform/fpl/model/Recipients.java (hunk 9 above)
Comment (akomisarek) [inline service/src/main/java/uk/gov/hmcts/reform/fpl/model/Recipients.java]: We probably made a mistake by allowing the screens to capture the data in different way :(.

File: ccd-definition/AuthorisationCaseField/solicitor.json [hunk 17]
```diff
@@ -292,5 +292,33 @@
...
+    "CaseFieldID": "statementOfService",
+    "UserRole": "caseworker-publiclaw-solicitor",
+    "CRUD": "CRUD"
+  },
+  {
+    "LiveFrom": "01/01/2017",
+    "CaseTypeID": "CARE_SUPERVISION_EPO",
+    "CaseFieldID": "serviceDeclarationLabel",
+    "UserRole": "caseworker-publiclaw-solicitor",
+    "CRUD": "CRU"
```
[the remark this answers is not in the document]
Comment (rebeccabrennan) [reply ccd-definition/AuthorisationCaseField/solicitor.json]: No we do not :)

File: e2e/tests/localAuthorityMaintainsCaseAfterSubmission_test.js (hunk 15 above)
Comment (rebeccabrennan) [reply e2e/tests/localAuthorityMaintainsCaseAfterSubmission_test.js]: Yes they do, without them it doesnt give the page enough time to load and it just skips entering the form completely.

File: service/src/main/java/uk/gov/hmcts/reform/fpl/model/CaseData.java (hunk 16 above)
Comment (rebeccabrennan) [reply service/src/main/java/uk/gov/hmcts/reform/fpl/model/CaseData.java]: i agree I think it should be a List of recipients

File: ccd-definition/CaseEventToFields/addStatementOfService.json [hunk 18]
```diff
@@ -0,0 +1,47 @@
...
+    "DisplayContext": "READONLY",
+    "PageID": 1,
+    "PageDisplayOrder": 1,
+    "PageColumnNumber": 1
+  },
+  {
+    "LiveFrom": "01/01/2017",
+    "CaseTypeID": "CARE_SUPERVISION_EPO",
+    "CaseEventID": "addStatementOfService",
+    "CaseFieldID": "serviceConsent",
```
[the remark this answers is not in the document]
Comment (rebeccabrennan) [reply ccd-definition/CaseEventToFields/addStatementOfService.json]: I dont think there is anywhere else I can put it?

File: e2e/tests/localAuthorityMaintainsCaseAfterSubmission_test.js (hunk 15 above)
Comment (akomisarek) [reply e2e/tests/localAuthorityMaintainsCaseAfterSubmission_test.js]: Ok interesting. Thanks.

**Pull request merged**